server/index.ts

```diff
@@ -189,6 +193,69 @@ router.get('/:space/proposal/:id/votes', async (req, res) => {
     .then(obj => res.json(obj));
 });
 
+router.post('/:space/offchain_proofs', async (req, res) => {
+  const space = req.params.space;
+  const verifyingContract = req.body.verifyingContract;
+  const actionId = req.body.actionId;
+  const chainId = req.body.chainId;
+  const merkleRoot = req.body.merkleRoot;
+  const steps = req.body.steps;
+
+  if (
+    !space ||
+    !verifyingContract ||
+    !actionId ||
+    !chainId ||
+    !merkleRoot ||
+    !steps ||
+    steps.length < 1 // must have at least 2 steps
+  )
+    return res.status(400).send({
+      error: 'invalid request parameters'
+    });
+
+  const indexes = steps.map(s => s.index).filter(s => s.index);
+  if (indexes.length !== steps.length) {
+    return res.status(400).send({ error: 'invalid merkle root' });
+  }
+
+  // the initial index should be always 0
+  let current = indexes[0];
+  if (current !== 0)
+    return res.status(400).send({ error: 'invalid initial index' });
+
+  // each step must have the index provided in an incremental order by 1 unit only
+  for (let i = 1; i < indexes.length; i++) {
+    if (indexes[i] - current !== 1) {
+      return res.status(400).send({ error: 'invalid indexes' });
+    }
+    current = indexes[i];
+  }
+
+  const merkleTree = new MerkleTree(
+    steps.map(s =>
+      buildVoteLeafHashForMerkleTree(s, verifyingContract, actionId, chainId)
+    )
+  );
+
+  if (merkleTree.getHexRoot() === merkleRoot) {
+    await saveOffchainProof(space, merkleRoot, steps);
+    return res.status(201).send();
+  }
+
+  return res.status(403).send({ error: 'invalid merkle root' });
+});
+
+router.get('/:space/offchain_proof/:merkleRoot', async (req, res) => {
+  const { space, merkleRoot } = req.params;
+  return getOffchainProof(space, merkleRoot)
+    .then(p => res.status(200).send(p))
+    .catch(e => {
+      console.error(e);
+      return res.status(404).send();
```

Comment:
Perhaps this should be a `500`, and inside the `.then(p => ...)` if the the `p.length === 0` (I think it's an `Array`?) then it should `res.status(404).send()` then `return`?

Reply:
Yeah, it is an array. Just changed it accordingly.